sql/core/src/main/scala/org/apache/spark/sql/execution/stat/StatFunctions.scala

```diff
@@ -77,4 +78,42 @@ private[sql] object StatFunctions {
       })
     counts.cov
   }
+
+  /** Generate a table of frequencies for the elements of two columns. */
+  private[sql] def crossTabulate(df: DataFrame, col1: String, col2: String): DataFrame = {
+    val tableName = s"${col1}_$col2"
+    val distinctCol2 = df.select(col2).distinct.orderBy(col2).collect()
```

Comment:
The first implementation uses multiple passes. We need two passes any way, either on the original columns or on the pair counts. The latter may be better.

1. `val counts = select(col1, col2).rdd.countByValue().cache()`.
2. Get distinct values from col2: `counts.map(_._1._2).distinct().collect()`. I'm not sure whether ordering by counts is useful here.
3. GroupBy `col1` in `counts` and create an RDD of Row. And then apply the schema.